<a href="https://colab.research.google.com/github/alexsasik/for-dz/blob/main/online_sex_work.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Исследовательская задача**

Датасет, выбранный для исследования, представляет собой большое количество данных с форума для взрослых. Однако сами создатели написали в своей работе, что основная исследовательская задача не выполнена. Только 78 исследуемым был рассчитан риск их вовлечения в оказание сексуальных услуг онлайн. 

Следовательно, нашей основной задачей становится выявление факторов, влияющих на риск, на основе 78 уже исследованных людей и расчет риска для неисследованных. Кроме того, датасет был с технической ошибкой, вследствие которой после собранных данных шло огромное количество пустых мест. Поэтому для начала необходимо убрать их, и начать работать уже с чистым датасетом.

## **Переменные:**

*   **User_ID** - идентификационный номер пользователя на форуме
*   **Gender** - пол пользователя, male/female
*   **Age** - возраст пользователя в годах
*   **Location** - Местоположение пользователя, в целях конфиденциальности показано как название точки (A, B и т.д.)
*   **Verification** - является пользователь верифицированнным на форуме или нет
*   **Sexual_orientation** - сексуальная ориентация пользователя (гетеро-, би- или гомосексуальная)
*   **Sexual_polarity** - сексуальная полярность пользователя
*   **Looking_for** - предпочтения пользователя в выборе пары
*   **Points_Rank** - количество очков на форуме
*   **Last_login** - время последнего пребывания в сети с момента обращения к статистике форума
*   **Member_since** - дата в формате ДД.ММ.ГГГГ, в которую пользователь зарегистрировался на форуме
*   **Number_of_comments_in_public_forum** - число комментариев, оставленных пользователем
*   **Time_spent_chatting_H:M** - время, проведенное за общением на форуме в формате Часы:Минуты
*   **Number_of_adbertisments_posted** - количество размещенных рекламных сообщений
*   **Number_of_offline_meetings_attended** - количество посещенных оффлайн-свиданий после общения на форуме
*   **Profile_pictures** - количество фотографий в профиле пользователя
*   **Friends_ID_list** - список идентификационных номеров друзей пользователя с разделителем в виде запятой
*   **Risk** - риск быть вовлеченным в предоставление сексуальных услуг онлайн











## **Очистка датасета от пустых значений**

In [ ]:
import pandas as pd

In [ ]:
data = pd.read_csv('https://github.com/alexsasik/for-dz/raw/main/online_sex_work.csv')
data

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (1,2,3,4,5,6,7,8,9,10,11,12,16,17) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


User_ID Gender   Age Location  Verification Sexual_orientation  \
0        10386.0   male  34,6        A  Non_Verified         Homosexual   
1           14.0   male  32,2        J  Non_Verified       Heterosexual   
2        16721.0   male  33,6        K  Non_Verified       Heterosexual   
3        16957.0   male    34        H  Non_Verified       Heterosexual   
4        17125.0   male  39,5        B  Non_Verified       Heterosexual   
...          ...    ...   ...      ...           ...                ...   
1048570      NaN    NaN   NaN      NaN           NaN                NaN   
1048571      NaN    NaN   NaN      NaN           NaN                NaN   
1048572      NaN    NaN   NaN      NaN           NaN                NaN   
1048573      NaN    NaN   NaN      NaN           NaN                NaN   
1048574      NaN    NaN   NaN      NaN           NaN                NaN   

        Sexual_polarity Looking_for Points_Rank      Last_login Member_since  \
0                Switch         Men          50  before_10_days    17.9.2012   
1              Dominant       Women         518   before_1_days    1.11.2009   
2              Dominant       Women         150   before_3_days     1.4.2013   
3              Dominant       Women         114   before_4_days     8.4.2013   
4              Dominant       Women         497   before_5_days    14.4.2013   
...                 ...         ...         ...             ...          ...   
1048570             NaN         NaN         NaN             NaN          NaN   
1048571             NaN         NaN         NaN             NaN          NaN   
1048572             NaN         NaN         NaN             NaN          NaN   
1048573             NaN         NaN         NaN             NaN          NaN   
1048574             NaN         NaN         NaN             NaN          NaN   

        Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
0                                       32                    0:2   
1                                      710                   3:45   
2                                       25                   2:15   
3                                      107                 359:22   
4                                      600                   0:21   
...                                    ...                    ...   
1048570                                NaN                    NaN   
1048571                                NaN                    NaN   
1048572                                NaN                    NaN   
1048573                                NaN                    NaN   
1048574                                NaN                    NaN   

         Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
0                                   0.0                                  0.0   
1                                   9.0                                  0.0   
2                                   1.0                                  1.0   
3                                   1.0                                  0.0   
4                                   0.0                                  6.0   
...                                 ...                                  ...   
1048570                             NaN                                  NaN   
1048571                             NaN                                  NaN   
1048572                             NaN                                  NaN   
1048573                             NaN                                  NaN   
1048574                             NaN                                  NaN   

         Profile_pictures                                    Friends_ID_list  \
0                     0.0                                              18260   
1                     0.0          11778,32024,43768,23969,27318,45884,31277   
2                    45.0                                  19805,21721,19802   
3                     1.0  40847,3818

Необходимо убрать все пустые строки. Такими будут те, где не значится **User_ID**. Отсутствие ID означает отсутствие пользователя.

In [ ]:
firstnan = data[data['User_ID'].isnull()].iloc[0]
firstnan

User_ID                                NaN
Gender                                 NaN
Age                                    NaN
Location                               NaN
Verification                           NaN
Sexual_orientation                     NaN
Sexual_polarity                        NaN
Looking_for                            NaN
Points_Rank                            NaN
Last_login                             NaN
Member_since                           NaN
Number_of_Comments_in_public_forum     NaN
Time_spent_chating_H:M                 NaN
Number_of_advertisments_posted         NaN
Number_of_offline_meetings_attended    NaN
Profile_pictures                       NaN
Friends_ID_list                        NaN
Risk                                   NaN
Name: 28831, dtype: object

Создадим **срез** датасета, который будет очищен от пустых значений и назовем его **cdata** (clean data). С этим срезом мы будем работать дальше, чтобы пустые значения не искажали статистику.

In [ ]:
cdata = data.iloc[:28831]
cdata

User_ID  Gender   Age Location  Verification Sexual_orientation  \
0      10386.0    male  34,6        A  Non_Verified         Homosexual   
1         14.0    male  32,2        J  Non_Verified       Heterosexual   
2      16721.0    male  33,6        K  Non_Verified       Heterosexual   
3      16957.0    male    34        H  Non_Verified       Heterosexual   
4      17125.0    male  39,5        B  Non_Verified       Heterosexual   
...        ...     ...   ...      ...           ...                ...   
28826   9962.0    male  27,2        J  Non_Verified       Heterosexual   
28827   9964.0    male  46,4        J  Non_Verified          bicurious   
28828   9966.0    male  28,8        C  Non_Verified       Heterosexual   
28829   9968.0    male  31,5        J  Non_Verified           bisexual   
28830    998.0  female  38,7        F  Non_Verified       Heterosexual   

      Sexual_polarity    Looking_for Points_Rank       Last_login  \
0              Switch            Men          50   before_10_days   
1            Dominant          Women         518    before_1_days   
2            Dominant          Women         150    before_3_days   
3            Dominant          Women         114    before_4_days   
4            Dominant          Women         497    before_5_days   
...               ...            ...         ...              ...   
28826          Switch          Women           0   before_23_days   
28827       Submisive  Men_and_Women          15  before_597_days   
28828       Submisive          Women          15    before_4_days   
28829       Submisive            Men          30  before_665_days   
28830        Dominant         Nobody          20  before_157_days   

      Member_since Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
0        17.9.2012                                 32                    0:2   
1        1.11.2009                                710                   3:45   
2         1.4.2013                                 25                   2:15   
3         8.4.2013                                107                 359:22   
4        14.4.2013                                600                   0:21   
...            ...                                ...                    ...   
28826     5.9.2012                                  0                    0:3   
28827     5.9.2012                                  0                    0:0   
28828     5.9.2012                                  0                    0:0   
28829     1.9.2012                                  0                   2:54   
28830    20.5.2010                                  0                    0:9   

       Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
0                                 0.0                                  0.0   
1                                 9.0                                  0.0   
2                                 1.0                                  1.0   
3                                 1.0                                  0.0   
4                                 0.0                                  6.0   
...                               ...                                  ...   
28826                             1.0                                  0.0   
28827                             0.0                                  0.0   
28828                             0.0                                  0.0   
28829                             4.0                                  0.0   
28830                             0.0                                  0.0   

       Profile_pictures                                    Friends_ID_list  \
0                   0.0                                              18260   
1                   0.0          11778,32024,43768,23969,27318,45884,31277   
2                  45.0                                  19805,21721,19802   
3                   1.0  40847,38183,9507,42259,5807,28118,24848,37170,...   
4

## **Частотные таблицы**

Создадим частотные таблицы для всех переменных, содержащих категориальные данные, кроме переменной **Risk**, так как риск по ней не рассчитан. Для переменных **Age, Points_Rank, Member_since, Number_of_Comments_in_public_forum, Time_spent_chating_H:M, Friends_ID_list**	количество значений будет слишком большое, из-за чего данные будет сложно интерпретировать. Из них данные можно будет извлечь в новые объединенные переменные (например, разбить возраст на промежутки в 10 лет). В то же время такие переменные, как **Last_login, Number_of_advertisments_posted, Number_of_offline_meetings_attended, Profile_pictures** не обязательно будут иметь слишком большой разброс значений, поэтому, для проверки этой гипотезы, мы создадим для них частотные таблицы.

Эти таблицы дадут нам понимание о преобладающих демографических показателях пользователей форума и их преобладающих сексуальных предпочтениях.

In [20]:
cdata['Gender'].value_counts(dropna=False)

male      22382
female     6445
NaN           4
Name: Gender, dtype: int64

Эта таблица показывает, что большая потребность в форуме для взрослых встречается у пользователей мужского пола, поскольку их почти в 3,5 раза больше, чем женщин.

In [21]:
cdata['Location'].value_counts(dropna=False)

A                     10944
B                      3572
G                      2211
L                      2210
M                      1261
K                      1209
C                      1176
H                      1167
E                       951
O                       892
J                       849
F                       822
N                       623
S                       484
D                       455
A kraj                    2
Jihomoravsk_ý kraj        1
NaN                       1
South Moravian            1
Name: Location, dtype: int64

Частотная таблица по локации пользователей показывает, что целевой аудиторией форума является один населенный пункт/область, так как почти треть пользователей находятся в локации А (в чистом датасете 28830 пользователей).

In [22]:
cdata['Verification'].value_counts(dropna=False)

Non_Verified     28228
Verified           602
Administrator        1
Name: Verification, dtype: int64

Из данной таблицы можно заключить, что подтвержденных пользователей довольно мало по сравнению с неподтвержденными. Кроме того, мы видим, что на форуме есть только 1 администратор.

In [23]:
cdata['Sexual_orientation'].value_counts(dropna=False)

Heterosexual    20441
bisexual         4391
bicurious        2992
Homosexual       1006
NaN                 1
Name: Sexual_orientation, dtype: int64

Среди пользователей форума примерно две трети являются гетеросексуальными, а их соотношение с гомосексуальными примерно равно 20:1.

In [24]:
cdata['Sexual_polarity'].value_counts(dropna=False)

Submisive    10878
Dominant     10474
Switch        7478
NaN              1
Name: Sexual_polarity, dtype: int64

Количество пользователей, проявляющих только доминантную или только подчиняющуюся сексуальную активность, примерно равно и составляет чуть больше двух третей, тогда как оставшуюся неполную треть занимают пользователи, у которых проявление сексуальной активности непостоянно.

In [25]:
cdata['Looking_for'].value_counts(dropna=False)

Nobody              19928
Women                4896
Nobody_but_maybe     1512
Men_and_Women        1074
Men                   996
NaN                   425
Name: Looking_for, dtype: int64

Из частотной таблицы можно заключить, что большая часть пользователей находится на форуме скорее не для знакомств, а для общения или чтения дискуссий. Наиболее же часто встречающийся пол искомого партнера - женский, что предсказуемо, так как большую часть аудитории форума составляют гетеросексуальные мужчины.

In [26]:
cdata['Last_login'].value_counts(dropna=False)

before_1_days       1392
before_3_days        961
before_4_days        314
before_6_days        237
before_5_days        233
                    ... 
before_2113_days       1
before_2376_days       1
before_2097_days       1
before_2026_days       1
before_2398_days       1
Name: Last_login, Length: 2360, dtype: int64

Среди топ-5 дат последнего онлайна все даты вписываются в промежуток недели до момента создания датасета. В то же время количество пользователей, которые были в сети хотя бы раз за неделю до сбора данных, составляет примерно 9-10% (мы пропускаем тех, кто был за 2 и за 7 дней до сбора данных в сети, так как это не очень важно). Это дает нам понимание, что примерно такой процент от всей аудитории форума достаточно активен.

In [28]:
cdata['Number_of_advertisments_posted'].value_counts(dropna=False)

0.0     25164
1.0      2984
2.0       453
3.0       114
4.0        48
5.0        28
6.0        12
8.0         6
7.0         5
10.0        3
9.0         3
11.0        3
14.0        2
18.0        2
16.0        1
15.0        1
17.0        1
21.0        1
Name: Number_of_advertisments_posted, dtype: int64

Подавляющее большинство пользователей форума ни разу не публиковало рекламного сообщения, а количество тех, кто публиковал хотя бы одно, меньше примерно в 9-10 раз. Дважды публиковало сообщения уже в 5 раз меньшее количество, чем единожды. 

In [29]:
cdata['Number_of_offline_meetings_attended'].value_counts(dropna=False)

0.0     28302
1.0       293
2.0        91
3.0        50
4.0        34
5.0        21
6.0        11
7.0         9
9.0         6
10.0        4
11.0        3
8.0         3
13.0        1
16.0        1
23.0        1
15.0        1
Name: Number_of_offline_meetings_attended, dtype: int64

По этой частотной таблице можно сказать, что пользователи в основном пользуются форумом для онлайн-общения. Более 98% ни разу не встречались со своими собеседниками в реальной жизни.

In [30]:
cdata['Profile_pictures'].value_counts(dropna=False)

0.0     26351
1.0       936
2.0       396
3.0       257
4.0       185
        ...  
67.0        1
53.0        1
45.0        1
34.0        1
40.0        1
Name: Profile_pictures, Length: 63, dtype: int64

По полученным данным можно судить о том, что пользователи форума для взрослых не любят публиковать свои личные данные. 91,5% пользователей не устанавливал фотографию профиля и сохранял визуальную анонимность.

## **Меры центральной тенденции**

Рассчитаем меры центральной тенденции для основных переменных:

In [ ]:
cdata.columns

Index(['User_ID', 'Gender', 'Age', 'Location', 'Verification',
       'Sexual_orientation', 'Sexual_polarity', 'Looking_for', 'Points_Rank',
       'Last_login', 'Member_since', 'Number_of_Comments_in_public_forum',
       'Time_spent_chating_H:M', 'Number_of_advertisments_posted',
       'Number_of_offline_meetings_attended', 'Profile_pictures',
       'Friends_ID_list', 'Risk'],
      dtype='object')

In [ ]:
clmns = ['User_ID', 'Gender', 'Age', 'Location', 'Verification',
       'Sexual_orientation', 'Sexual_polarity', 'Looking_for', 'Points_Rank',
       'Last_login', 'Member_since', 'Number_of_Comments_in_public_forum',
       'Time_spent_chating_H:M', 'Number_of_advertisments_posted',
       'Number_of_offline_meetings_attended', 'Profile_pictures',
       'Friends_ID_list', 'Risk']

In [ ]:
cdata['User_ID'].mode()

0      2035.0
1      2589.0
2      2773.0
3      3212.0
4      4349.0
       ...   
63    48026.0
64    48232.0
65    48326.0
66    48438.0
67    48490.0
Length: 68, dtype: float64